### ISSUE TITLE
Implement Cosmos block manager

### ISSUE THREAD

**Issue by michaelfig** [Agoric/agoric-sdk#667]
This design is to protect the SwingSet kernel from having to deal with the rewinding, then recalculation of mailbox and kernel state that can happen when called via the Cosmos SDK.

# The Problem

We were naive about what messages should be passed from Cosmos SDK to the SwingSet kernel, relying on the kernel's deduplication of inbound SwingSet messages in order to succeed.

The latest Tendermint (v0.33.0)[https://github.com/tendermint/tendermint/tree/v0.33.0] apparently has the following behaviour:

1. The block proposer creates Block N with its transactions and is asked to produce its state.
2. The Cosmos SDK calls:
  a. BeginBlock (dispatched to modules in SetOrderBeginBlockers)
  b. Transactions in block N (dispatched to individual modules, such as x/swingset)
  c. EndBlock (dispatched to modules in SetOrderEndBlockers)
3. Any keeper changes that were made in the above step are used to calculate Block N's state (app hash).
4. The block proposer signs the resulting state and gossips it to the other validators.
5. Each of the participating validators (**including the block proposer** (new in Tendermint v0.33.0)) are given Block N and are asked to produce the state.
6. The same calls as in steps 2 and 3 are made.
7. The consensus algorithm works to ensure agreement on a given state.  It may repeat from step 1 for the same Block N, but with different transactions.
8. Block N is agreed on and committed, so we repeat from step 1.

Cosmos SDK expects SwingSet to implement a pure function from N + transactions to kernel state.

The main problem vis-a-vis the SwingSet kernel is that the kernel is not a pure function since it incrementally mutates the Javascript heap and we have no efficient way of rewinding state to Block N-1.

# Solution Stages

There are three stages to the solution:

1. Current behaviour: ad-hoc attempts to save state and use kernel message deduplication.  Does not work for the timer device, since it is not idempotent in regards to rerunning the timer with the same timestamp.
2. Accommodate rerunning exactly the same transactions if Block N is repeated (solves only the double-run of the block proposer in steps 2 and 6)
3. Accommodate efficient rewinding of kernel state to Block N-1 (solves the whole issue, but requires snapshots of kernel state)

# Solution Details

Introduce an x/swingset/handler.go handler for EndBlock, which sends an `END_BLOCK` controller call to ag-chain-cosmos.

Introduce an ag-chain-cosmos-specific "Block Manager", which records all SwingSet controller calls (`BEGIN_BLOCK`, `DELIVER_INBOUND`, `END_BLOCK`) from x/swingset/handler.

The Block Manager uses this algorithm (more or less):

```js
let currentMsg;
let currentMessages = [];
let currentHeight;
let replaying = true;
let lastMessages = [];
let lastHeight = 0;
let lastKeeperState;

let newMessages;
let newKeeperState;
function blockManager(msg) {
  if (msg.type === 'BEGIN_BLOCK') {
    // Start a new block.
    replaying = true;
    currentMsg = 0;
    currentMessages = [];
    currentHeight = msg.blockHeight;
  } else {
    // We're working on the next message of the same block.
    currentMsg += 1;
  }

  // Keep track of what we saw.
  currentMessages.push(msg);

  if (!replaying) {
    // Compute new state by running the kernel.
    newKeeperState = kernelPerformMessage(msg);
  } else if (!messagesAreEqual(msg, lastMessages[currentMsg])) {
    // Divergence of the inbound messages, so rewind state if we need to.
    if (currentHeight - 1 !== lastHeight) {
      // For solution stage 2, this call would always throw.
      newKeeperState = setKernelState(currentHeight - 1);
    }
    replaying = false;
    // Resend the messages we have so far to the restored kernel.
    for (const m of currentMessages) {
      newKeeperState = kernelPerformMessage(m);
    }
  }

  if (msg.type === 'END_BLOCK') {
    // Commit all keeper state, even on replay.
    keeperCommit(newKeeperState);

    // Commit the kernel state but only if we didn't replay.
    if (!replaying) {
      kernelCommitState();
    }

    // Advance our stored state variables.
    lastHeight = currentHeight;
    lastMessages = currentMessages;
    lastKeeperState = newKeeperState;
  }
}
```

**Comment by michaelfig** [comment]
@warner PTAL when you have time.

**Comment by michaelfig** [comment]
I've refined the solution algorithm and expressed it as pseudocode...it should be easier to understand now.

**Comment by michaelfig** [comment]
We can do even better than this.  Adding a `COMMIT_BLOCK` event can almost ensure that we don't commit kernel state unless the block has been committed, too.

Implemented in #669, see the PR for algorithm details (especially `block-manager.js`).

**Comment by michaelfig** [comment]
Oh, and #669 panics if the kernel state cannot be rewound.  This is crash-only behaviour, which means a rewind attempt will restart the chain from the last committed state, which should be all we need to clumsily support multiple different proposed blocks at the same height.  Eventually, we will hook in with the proper rewinding support to do this more efficiently (and not panic).
